crates/ironrdp-pdu/src/geometry.rs

```diff
@@ -1,90 +1,202 @@
 use std::cmp::{max, min};
-use std::io;
 
-use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use crate::cursor::ReadCursor;
+use crate::{PduDecode, PduEncode, PduResult};
+
+pub(crate) mod private {
+    pub struct BaseRectangle {
+        pub left: u16,
+        pub top: u16,
+        pub right: u16,
+        pub bottom: u16,
+    }
+
+    impl BaseRectangle {
+        pub fn empty() -> Self {
+            Self {
+                left: 0,
+                top: 0,
+                right: 0,
+                bottom: 0,
+            }
+        }
+    }
+
+    pub trait RectangleImpl: Sized {
+        fn from_base(rect: BaseRectangle) -> Self;
+        fn to_base(&self) -> BaseRectangle;
+
+        fn left(&self) -> u16;
+        fn top(&self) -> u16;
+        fn right(&self) -> u16;
+        fn bottom(&self) -> u16;
+    }
+}
+
+use private::*;
+
+pub trait Rectangle: RectangleImpl {
+    fn width(&self) -> u16;
+    fn height(&self) -> u16;
+
+    fn empty() -> Self {
+        Self::from_base(BaseRectangle::empty())
+    }
+
+    fn union_all(rectangles: &[Self]) -> Self {
+        Self::from_base(BaseRectangle {
+            left: rectangles.iter().map(|r| r.left()).min().unwrap_or(0),
+            top: rectangles.iter().map(|r| r.top()).min().unwrap_or(0),
+            right: rectangles.iter().map(|r| r.right()).max().unwrap_or(0),
+            bottom: rectangles.iter().map(|r| r.bottom()).max().unwrap_or(0),
+        })
+    }
+
+    fn intersect(&self, other: &Self) -> Option<Self> {
+        let a = self.to_base();
+        let b = other.to_base();
+
+        let result = BaseRectangle {
+            left: max(a.left, b.left),
+            top: max(a.top, b.top),
+            right: min(a.right, b.right),
+            bottom: min(a.bottom, b.bottom),
+        };
+
+        if result.left < result.right && result.top < result.bottom {
+            Some(Self::from_base(result))
+        } else {
+            None
+        }
+    }
+
+    fn union(&self, other: &Self) -> Self {
+        let a = self.to_base();
+        let b = other.to_base();
+
+        let result = BaseRectangle {
+            left: min(a.left, b.left),
+            top: min(a.top, b.top),
+            right: max(a.right, b.right),
+            bottom: max(a.bottom, b.bottom),
+        };
+
+        Self::from_base(result)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct InclusiveRectangle {
+    pub left: u16,
+    pub top: u16,
+    pub right: u16,
+    pub bottom: u16,
+}
 
-/// An **inclusive** rectangle.
-///
-/// This struct is defined as an **inclusive** rectangle.
-/// That is, the pixel at coordinate (right, bottom) is included in the rectangle.
```

Comment:
Let’s keep this doc comment for both `InclusiveRectangle` and `ExclusiveRectangle` (changing it a bit for `ExclusiveRectangle` since it’s exclusive)

Reply:
Sure! Thinks, it just slipped through the cracks while I was investigating a better approach to handle rectangles!